certstore/certstore.go

```diff
@@ -287,13 +287,31 @@ func (cs *Store) GetPowerTable(ctx context.Context, instance uint64) (gpbft.Powe
 		return nil, fmt.Errorf("cannot return a power table before the first instance: %d", cs.firstInstance)
 	}
 
-	lastPowerTable := cs.firstInstance
-	if latestCert := cs.Latest(); latestCert != nil {
-		lastPowerTable = latestCert.GPBFTInstance + 1
-	}
-	if instance > lastPowerTable {
-		return nil, fmt.Errorf("cannot return future power table for instance %d > %d", instance, lastPowerTable)
-	}
+	// Copy a reference to both the latest cert and latest power table while holding
+	// the lock to guarantee order in case the latest changes while the requested
+	// power table is being retrieved.
+	cs.mu.RLock()
+	latestCert := cs.latestCertificate
+	latestPowerTable := cs.latestPowerTable
+	cs.mu.RUnlock()
+
+	nextCertInstance := cs.firstInstance
+	if latestCert != nil {
+		nextCertInstance = latestCert.GPBFTInstance + 1
+	}
+	if instance > nextCertInstance {
+		return nil, fmt.Errorf("cannot return future power table for instance %d > %d", instance, nextCertInstance)
+	}
+	if instance == nextCertInstance && len(latestPowerTable) != 0 {
```

Comment:
Instead of checking `len(latestPowerTable)`, can we check `latestCert != nil`? I guess that won't work for instance 0... meh, this works fine.

Reply:
Yes, it would be problematic for `firstInstance`. An alternative is to explicitly check if the `instance` is the same as `firstInstance`. But that makes up a more complicated logic.